Recreate passing tests and file-based and in-memory runQuery

paweljakubas opened the pull request:
# Issue Number

<!-- Put here a reference to the issue this PR relates to and which requirements it tackles -->
#154 


# Overview

<!-- Detail in a few bullet points the work accomplished in this PR -->

- [x] I have added unit tests that check file-backed sqlite db read/write/clean operation in a number of state changing scenerios
- [x] I have added property test that checks if we are in line with in-memory version when using random number of state changing operations in a row 


# Comments

<!-- Additional comments or screenshots to attach if any -->
This PR checks DB in two modes, ie., two places where db can reside : in-memory and in the file. And it seems they are not the same addressed within persistent, hence different `runQuery'`. Secondly, because these are two completely different mechanism (memory vs file system) they are differently serviced when detaching is realized. in-memory detaches gracefully, but `file-based` solution not so gracefully. It turn out that when we get rid of `close` we can easily end up with situation where we have too many file handles open and :
 ```
uncaught exception: SqliteException
SQLite3 returned ErrorCan'tOpen while attempting to perform open "backup/test.db".
(after 504 tests and 4 shrinks)
```
As evidenced above this happened after 503 successful test of  `prop_randomOpChunks` (I commented `close db` and added `not (null pairs) ==> withMaxSuccess 1000 $  monadicIO (pure inMemoryDB >>= prop)`
When `close db` is on the test passes. This was the reason I decided to spend time investigating this issue. This also suggests to me that in `file-based mode` we need to use `close` and the test is needed, also `runQuery'` workaround. Otherwise, someone with small number of open handles in his/her computer will be subjected to this limitation, as anyone that uses our wallet-backend without interruption for long time (I am thinking here about exchanges). 
<!-- 
Don't forget to:

 ✓ Self-review your changes to make sure nothing unexpected slipped through
 ✓ Assign yourself to the PR
 ✓ Assign one or several reviewer(s)
 ✓ Once created, link this PR to its corresponding ticket
 ✓ Acknowledge any changes required to the Wiki
-->

paweljakubas committed:
Recreate passing tests and file-based and in-memory runQuery

paweljakubas commented on lib/core/backup/.gitignore:
```diff
@@ -0,0 +1,4 @@
+# Ignore everything in this directory
+*
+# Except this file
+!.gitignore
```
here will come mechanism of dealing with directories as in DB bench PR

paweljakubas commented on lib/core/src/Cardano/Wallet/DB/Sqlite.hs:
```diff
@@ -139,13 +141,21 @@ import qualified Database.Sqlite as Sqlite
+enableForeignKeys conn = do
+    stmt <- Sqlite.prepare conn "PRAGMA foreign_keys = ON;"
+    _ <- Sqlite.step stmt
+    Sqlite.finalize stmt
```
this is needed for file-based. otherwise :  
```
      1) Cardano.Wallet.DB.SqliteCorruption, Check db opening/closing, opening and closing of db works
           uncaught exception: SqliteException
           SQLite3 returned ErrorBusy while attempting to perform close: unable to close due to unfinalized statements or unfinished backups
```

paweljakubas commented on lib/core/src/Cardano/Wallet/DB/Sqlite.hs:
```diff
@@ -139,13 +141,21 @@ import qualified Database.Sqlite as Sqlite
+    pure (conn, backend)
+
+createSqliteBackend1 :: Maybe FilePath -> LogFunc -> IO SqlBackend
+createSqliteBackend1 fp logFunc = do
```
this one can be married with `createSqliteBackend`

paweljakubas commented on lib/core/src/Cardano/Wallet/DB/Sqlite.hs:
```diff
@@ -176,29 +186,43 @@ handleConstraint e = handleJust select handler . fmap Right
-        runQuery' = withMVar bigLock . const . runQuery conn
+        runQuery' cmd = case fp of
+            Nothing ->
+                withMVar bigLock $ const $ runQuery backend cmd
```
in-memory - this cannot be used with file-based

paweljakubas commented on lib/core/src/Cardano/Wallet/DB/Sqlite.hs:
```diff
@@ -176,29 +186,43 @@ handleConstraint e = handleJust select handler . fmap Right
+            Nothing ->
+                withMVar bigLock $ const $ runQuery backend cmd
+            _ ->
+                withMVar bigLock $ const $ runResourceT $ runNoLoggingT $ withSqlConn (createSqliteBackend1 fp)
```
file-based, this cannot be used with in-memory. Needs other pairs of eyes - at this moment do not see why

paweljakubas commented on lib/core/test/unit/Cardano/Wallet/DB/SqliteFileModeSpec.hs:
```diff
@@ -0,0 +1,284 @@
+
+        verify dbM dbF
+
+        close conn
```
comment this and add `withMaxSuccess 1000` and will probably end up with : 
```
     uncaught exception: SqliteException
       SQLite3 returned ErrorCan'tOpen while attempting to perform open "backup/test.db".
       (after 504 tests and 4 shrinks)
```
as I did in my laptop.

It suggests we need closing when using file-based Sqlite

paweljakubas commented on lib/core/test/unit/Cardano/Wallet/DB/SqliteFileModeSpec.hs:
```diff
@@ -0,0 +1,284 @@
+    handleChunks chunk = do
+        (conn, db) <- fileDBLayer
+        forM_ chunk (updateDB db)
+        close conn
```
comment this also to encounter open handle sooner

jonathanknowles commented on lib/core/test/unit/Cardano/Wallet/DB/SqliteFileModeSpec.hs:
```diff
@@ -0,0 +1,284 @@
+prop_randomOpChunks
+    :: DBLayer IO (SeqState DummyTarget) DummyTarget
+    -> KeyValPairs (PrimaryKey WalletId) (Wallet (SeqState DummyTarget) DummyTarget , WalletMetadata)
+    -> Property
```
I would suggest the following:
```hs
prop_randomOpChunks
    :: DBLayer IO (SeqState DummyTarget) DummyTarget
    -> KeyValPairs
        (PrimaryKey WalletId)
        (Wallet (SeqState DummyTarget) DummyTarget, WalletMetadata)
    -> Property
```
This way, we can still satisfy https://github.com/input-output-hk/cardano-wallet/wiki/Coding-Standards#limit-line-length-to-80-characters

jonathanknowles commented on lib/core/test/unit/Cardano/Wallet/DB/SqliteFileModeSpec.hs:
```diff
@@ -0,0 +1,284 @@
+                Nothing )
+
+    withDB inMemoryDBLayer $
+        describe "random operation chunks property when writing to/reading from file" $ do
```
I'd suggest the following:

```hs
    withDB inMemoryDBLayer $
        describe
            "random operation chunks property when writing to/reading from \
            \file" $ do
```

For compliance with https://github.com/input-output-hk/cardano-wallet/wiki/Coding-Standards#limit-line-length-to-80-characters

KtorZ committed:
export 'SqlBackend' instead of 'SqlConnection'

This solves the problem when dealing with the SQLite in-memory vs in-file database. By using 'SqlBackend' and calling close',
we do not break the abstraction boundary set by persistent and leave the connection management entirely up to persistent

KtorZ committed:
refactor SqliteFileModeSpec to enhance readability & use close'

KtorZ committed:
remove unused 'DummyState' and its 'PersistState' instance

KtorZ reviewed (approved):
LGTM. I've added a couple of changes + refactorings on top, so waiting for @paweljakubas to review back.
